presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,191 @@
+    {
+        requireNonNull(allQueries);
+        requireNonNull(queryMemoryReservations);
+
+        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
+                .stream()
+                .filter(Objects::nonNull)
+                .filter(entry -> entry.getValue() > 0)
+                .filter(entry -> shouldClaimQueryAsLeaked(allQueries, entry.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (!leakedQueryReservations.isEmpty()) {
+            log.warn("Memory leak detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+        }
+
+        // keep numberOfLeakedQueries and leakedQueries in sync
+        synchronized (this) {
+            numberOfLeakedQueries.set(leakedQueryReservations.size());
+            leakedQueries.update(leakedQueryReservations.keySet());
+        }
+    }
+
+    private static boolean shouldClaimQueryAsLeaked(List<QueryInfo> allQueries, QueryId queryId)
+    {
+        boolean queryExists = false;
+        QueryState queryState = null;
+        DateTime queryEndTime = null;
+
+        for (QueryInfo queryInfo : allQueries) {
+            if (queryInfo.getQueryId().equals(queryId)) {
+                queryExists = true;
+                queryState = queryInfo.getState();
+                queryEndTime = queryInfo.getQueryStats().getEndTime();
+            }
+        }
+
+        if (!queryExists) {
+            return true;
+        }
+
+        if (queryState.equals(RUNNING) || queryEndTime == null) {
+            return false;
+        }
+
+        return secondsBetween(queryEndTime, now()).getSeconds() >= DEFAULT_LEAK_CLAIM_DELTA_SEC;
+    }
+
+    synchronized boolean isQueryLeaked(QueryId queryId)
+    {
+        return leakedQueries.contains(queryId);
+    }
+
+    int getNumberOfLeakedQueries()
+    {
+        return numberOfLeakedQueries.get();
+    }
+
+    interface LeakedQueries
+    {
+        boolean contains(QueryId queryId);
+        void update(Collection<QueryId> leakedQueries);
+    }
+
+    static class DefaultLeakedQueries
+            implements LeakedQueries
+    {
+        private final Map<QueryId, Boolean> leakedQueries;
+
+        DefaultLeakedQueries()
+        {
+            this.leakedQueries = new WeakHashMap<>();
+        }
+
+        public boolean contains(QueryId queryId)
+        {
+            return leakedQueries.containsKey(queryId);
+        }
+
+        @Override
+        public void update(Collection<QueryId> leakedQueries)
+        {
+            this.leakedQueries.clear();
+            leakedQueries.forEach(queryId -> this.leakedQueries.put(queryId, TRUE));
+        }
+    }
+
+    // For testing we need a list of strong references, otherwise the test will be non-deterministic due to GC effects.
+    static class TestLeakedQueries
```

Comment:
`DefaultLeakedQueries::update` clears the collection, so it could hold strong references without harm. Do we need two implementations?

Reply:
I am going to have a single impl. and simplify this.